### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by SanjoDeundiak**

### CODE REVIEW

File: Passw0rd/Protocol.cs [hunk 1]
```diff
@@ -64,162 +62,162 @@ public class Protocol
...
+                pheResp.C1.ToByteArray());
+            if (!isValid)
+            {
+                throw new ProofOfSuccessNotValidException();
+            }
 
-            var enrollment = model.Enrollment;
-            var nS = enrollment.Nonce;
+            var nS = pheResp.Ns;
             var nC = this.ctx.Crypto.GenerateNonce();
```
Comment (SanjoDeundiak) [inline Passw0rd/Protocol.cs]: Crypto operations should be encapsulated inside class. For example pheClient with interface like here https://github.com/VirgilSecurity/virgil-crypto-c/blob/master/codegen/models/project_phe/class_phe_client.xml

File: Passw0rd/Phe/SHA512.cs [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+namespace Passw0rd.Phe
+{
+    using System;
+    using System.IO;
+
+    using Org.BouncyCastle.Crypto.Digests;
+
+    public class SHA512
```
Comment (unlim-it) [inline Passw0rd/Phe/SHA512.cs]: Rename class name, it looks more like helper for SHA512

File: Passw0rd/ProtocolContext.cs [hunk 3]
```diff
@@ -69,131 +67,84 @@ private ProtocolContext()
...
-                var updateTokenList = updateTokens.Select(UpdateToken.Decode)
-                    .Where(it => it.Version > skCVer)
-                    .OrderBy(it => it.Version)
-                    .ToList();
-
-                ctx.UpdateTokens = updateTokenList;
-
-                foreach (var token in updateTokenList)
+                ctx.VersionedUpdateToken = StringUpdateTokenParser.Parse(updateToken);
+                if (ctx.VersionedUpdateToken.Version != pkSVer)
```
Comment (SanjoDeundiak) [inline Passw0rd/ProtocolContext.cs]: This line can be deleted

File: Passw0rd/Utils/StringKeyParser.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            {
+                throw new WrongServiceKeyException(e.ToString());
+            }
+            return (version, publicKey);
+        }
+
+        public (uint, SecretKey) ParseSecretKey(string clientSecretKey)
+        {
+            Validation.NotNullOrWhiteSpace(clientSecretKey);
+            var (version, keyBytes) = ParseKeyBytesByFlag(clientSecretKey, SecretKeyFlag);
```
Comment (Scratch-net) [inline Passw0rd/Utils/StringKeyParser.cs]: Prefix, not flag

File: Passw0rd/RecordUpdater.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        /// <returns>The updated Encrypted Passw0rd's record.</returns>
+        /// <param name="oldPwdRecord">Old Passw0rd's record.</param>
+        public byte[] Update(byte[] oldPwdRecord){
+            Validation.NotNullOrEmptyByteArray(oldPwdRecord, "Record isn't provided.");
+
+            var databaseRecord = DatabaseRecord.Parser.ParseFrom(oldPwdRecord);
+
+            if (databaseRecord.Version == VersionedUpdateToken.Version)
+            {
+                throw new WrongVersionException(
```
Comment (Scratch-net) [inline Passw0rd/RecordUpdater.cs:82]: We do not throw exception if record has the same version as the token version.

File: Passw0rd/Utils/StringKeyParser.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+ */
+
+namespace Passw0rd.Utils
+{
+    using System;
+    using Passw0rd.Phe;
+
+    internal class StringKeyParser
+    {
+        const string PublicKeyFlag = "PK";
```
Comment (Scratch-net) [inline Passw0rd/Utils/StringKeyParser.cs]: Prefix, not flag

**Review by Scratch-net — CHANGES_REQUESTED** [review]

File: Passw0rd/RecordUpdater.cs (hunk 5 above)
Comment (SanjoDeundiak) [reply Passw0rd/RecordUpdater.cs:82]: We do

**Pull request closed without merging**